### PR TITLE
EXPLAIN SQL doc

### LINKED ISSUE #430: sql: EXPLAIN statement
Document the `EXPLAIN` statement.

Some nuggets from Gitter:

kena @knz Jun 30 16:16
guys
who invented explain(debug)?
who can explain what it is supposed to do and what its output means?

RaduBerinde @RaduBerinde Jun 30 16:19
it is supposed to show the exact keys and values that are processed
i can answer any specific questions
the columns are rowid, key, value, and the last one indicates if the key corresponds to the end of a row, and if so what happens to that row (filtered, buffered, or a result)

kena @knz Jun 30 16:22
what do you mean with "processed"?
what does filtered mean? what does buffered mean?

RaduBerinde @RaduBerinde Jun 30 16:26
the SQL layer does KV ops and receveis KVs, right? then our code processes those KVs to form rows and whatnot
the debug output has an entry for each KV that our code looks at
filtered means that a row is thrown out because of a filter like WHERE
buffered means it is stored in memory in some node (e.g. sortNode). it willl show up again if that node outputs it

kena @knz Jun 30 16:28
aha
ok this becomes clearer now
but here you're talking about k/v reads
what about writes?

RaduBerinde @RaduBerinde Jun 30 16:32
reads. i have never used it on anything other than a SELECT, I am not sure if it even works

kena @knz Jun 30 16:32
yeah
ok

### PR DESCRIPTION
**Pull request by sploiselle**
Closes #430

Also includes some work from #794

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/1073)

### CODE REVIEW

**Comment by sploiselle** [comment]
@knz Thanks so much for your help with this. Can I get some final feedback on it?

File: explain.md [hunk 1]
```diff
@@ -18,6 +68,209 @@ The user requires the appropriate [privileges](privileges.html) for the statemen
 
 ## Parameters
 
-| Parameter | Description |
+Each `EXPLAIN` statement can only include one *Mode*, but can include any number of *Options*.
+
+| Parameter | Type | Description |
+|-----------|------|-------------|
+| `PLAN` | *Mode* | Return the query plan.<br/><br/>If you do not specify a mode, the statement uses `PLAN` mode. |
+| `TYPES` | *Mode* | Include the intermediate [data types](data-types.html) CockroachDB chooses to evaluate intermediate SQL expressions. <br/><br/>`TYPES` also implies `METADATA` and `EXPRS` options.|
```
Comment (knz) [inline explain.md]: TYPES is not a mode any more - it's now an option.


(The only remaining modes are PLAN, TRACE and DEBUG.)

File: explain.md [hunk 2]
```diff
@@ -18,6 +68,209 @@ The user requires the appropriate [privileges](privileges.html) for the statemen
 
 ## Parameters
 
-| Parameter | Description |
+Each `EXPLAIN` statement can only include one *Mode*, but can include any number of *Options*.
+
+| Parameter | Type | Description |
```
Comment (knz) [inline explain.md]: In the rendered table, the Type column is very narrow. Perhaps try to widen it in the markdown code by adding more dashes underneath its title.

**Comment by sploiselle** [comment]
@jseldess Ready for your edits

**Comment by jseldess** [comment]
This is great. Just a few suggestions for you, Sean. 

---

Review status: 0 of 3 files reviewed at latest revision, 10 unresolved discussions.

---

*[explain.md, line 15 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcU_2VfF_hIBONXaH-q:-KcU_2VfF_hIBONXaH-r:b-rfyo8n) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L15)):*
> ```Markdown
> Using `EXPLAIN`'s output, you can optimize your queries by taking the following points into consideration:
> 
> - Queries with fewer levels execute more quickly. Restructuring queries to require fewer levels of processing will generally improve performance.
> ```

Would you be willing to think up a simple example demonstrating such an optimization and add it to **Examples*, linking to that from here? 

---

*[explain.md, line 19 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUZvsFyEwtK6l6FYcT:-KcUZvsFyEwtK6l6FYcU:b-eyuxey) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L19)):*
> ```Markdown
> - Avoid scanning an entire table, which is the slowest way to access data. You can avoid this by [creating indexes](indexes.html) that contain at least one of the columns that the query is filtering in its `WHERE` clause.
> 
>   You can find out if your queries are performing entire table scans by using `EXPLAIN` to see which:
> ```

Do you think we could move this down into **Examples** and link to it from here?

---

*[explain.md, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUW2JABj9iCQ1k9oX8:-KcUW2JBfShflutmM5BL:b-gvf37w) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L75)):*
> ```Markdown
> | `EXPRS` | Include the SQL expressions that are involved in each processing stage. |
> | `QUALIFY` | Include table names when referencing columns, which might be important to verify the behavior of joins across tables with the same column names.<br/><br/>To list qualified names, `QUALIFY` requires you to include the `EXPRS` option. |
> | `METADATA` | Include the columns CockroachDB chooses to execute each processing step of the query in the **Columns** column of the output, as well as **Ordering** detail. |
> ```

I can't quite follow the `METADATA` description. Can you simplify the sentence, break it into two, or try to rephrase?

---

*[explain.md, line 89 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUWsn6pUcG784KjjL-:-KcUWsn6pUcG784KjjL0:b6iyone) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L89)):*
> ```Markdown
> |-----------|-------------|
> | **Level** | The level of hierarchy of the query plan. <br/><br/>`0` represents the last processing stage that produces the results sent to the SQL client receiving the data; the highest level represents the operation at the key-value layer accessing data. <br/><br/>The query plan has a tree structure; it is thus possible to see multiple processing stages at the same level, which indicates they are sibling stages feeding data to the previous processing stage with a lower level.|
> | **Type** | The query plan node's type, which are described in the [CockroachDB source on GitHub](https://github.com/cockroachdb/cockroach/pull/10055/files#diff-542aa8b21b245d1144c920577333ceedR764). |
> ```

We'll need to update this link once the PR has been merged, won't we?

---

*[explain.md, line 90 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUXPk90jYv_bLjnp64:-KcUXPkAW5xVzkHORqi9:burrt8x) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L90)):*
> ```Markdown
> | **Level** | The level of hierarchy of the query plan. <br/><br/>`0` represents the last processing stage that produces the results sent to the SQL client receiving the data; the highest level represents the operation at the key-value layer accessing data. <br/><br/>The query plan has a tree structure; it is thus possible to see multiple processing stages at the same level, which indicates they are sibling stages feeding data to the previous processing stage with a lower level.|
> | **Type** | The query plan node's type, which are described in the [CockroachDB source on GitHub](https://github.com/cockroachdb/cockroach/pull/10055/files#diff-542aa8b21b245d1144c920577333ceedR764). |
> | **Element** | The type of parameter being used by the query plan node. |
> ```

How do you get `Element` to show up in the response? I don't see it in any of the examples, and I haven't been able to get it in my own testing.

---

*[explain.md, line 121 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUYJuUjuDRtMoAcixz:-KcUYJuUjuDRtMoAciy-:bfgzkmp) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L121)):*
> ```Markdown
> 
> ~~~ sql
> > EXPLAIN (EXPRS) SELECT * FROM kv WHERE v > 3 order by v;
> ```

nit: `ORDER BY`

---

*[explain.md, line 141 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUYNqxbagWSQRyaW-H:-KcUYNqxbagWSQRyaW-I:bfgzkmp) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L141)):*
> ```Markdown
> 
> ~~~ sql
> > EXPLAIN (METADATA) SELECT * FROM kv WHERE v > 3 order by v;
> ```

nit: `ORDER BY`

---

*[explain.md, line 172 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUY5MdxbsxUm-gsKXK:-KcUY5MdxbsxUm-gsKXL:b95n5le) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L172)):*
> ```Markdown
> ~~~
> 
> {{site.data.alerts.callout_info}}In some cases the Ordering details report a column ordering with an equal sign (e.g. <code>=k</code>). This is a side effect of the internal ordering analysis performed by CockroachDB and merely indicates that CockroachDB has found that only one row matches a <code>WHERE</code> expression.{{site.data.alerts.end}}
> ```

nit: `In some cases, the <strong>Ordering</strong>` details...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-:-KcUZegAA34-unGd0IeF:bfmb83r)*

**Comment by jseldess** [comment]
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1073)*

**Comment by knz** [comment]
LGTM with the same nits/suggestions as Jesse.

Indeed, this is a rather nice page. I like it!

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[explain.md, line 89 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUWsn6pUcG784KjjL-:-KcYPptzh0FtfqFjtmLc:bnd42md) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L89)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

We'll need to update this link once the PR has been merged, won't we?
</blockquote></details>

Unfortunately it won't be merged before a month or two...

---

*[explain.md, line 90 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUXPk90jYv_bLjnp64:-KcYQ1WyAnDxIjM9GN74:b-4ybd0x) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L90)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

How do you get `Element` to show up in the response? I don't see it in any of the examples, and I haven't been able to get it in my own testing.
</blockquote></details>

Oh I think it's called "Field" now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-:-KcYQBBu5SSZhws6lbos:b-qvo16g)*

**Comment by sploiselle** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[explain.md, line 15 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcU_2VfF_hIBONXaH-q:-KcYgBPUQgyH2LgTRb3k:b-fjyqm2) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L15)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Would you be willing to think up a simple example demonstrating such an optimization and add it to **Examples*, linking to that from here? 
</blockquote></details>

@knz Can you help me in coming up with a seemingly reasonable query that can be optimized by understanding that the query planner splitting things into levels, and then what that optimization would be?

---

*[explain.md, line 19 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUZvsFyEwtK6l6FYcT:-KcYi-IO5BOFYdGyEuL0:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L19)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Do you think we could move this down into **Examples** and link to it from here?
</blockquote></details>

Done.

---

*[explain.md, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUW2JABj9iCQ1k9oX8:-KcYrZKf_Sl6C1itHqVk:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L75)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

I can't quite follow the `METADATA` description. Can you simplify the sentence, break it into two, or try to rephrase?
</blockquote></details>

Done.

---

*[explain.md, line 89 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUWsn6pUcG784KjjL-:-KcYreajhqY-reZK_NIX:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L89)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Unfortunately it won't be merged before a month or two...
</blockquote></details>

Done.

---

*[explain.md, line 90 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUXPk90jYv_bLjnp64:-KcYrgmc4TNHVTM44N-s:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L90)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Oh I think it's called "Field" now.
</blockquote></details>

Done.

---

*[explain.md, line 121 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUYJuUjuDRtMoAcixz:-KcYrkz5Kz85lkl_HKJj:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L121)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

nit: `ORDER BY`
</blockquote></details>

Done.

---

*[explain.md, line 141 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUYNqxbagWSQRyaW-H:-KcYrnEntyjdma4FNc8w:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L141)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

nit: `ORDER BY`
</blockquote></details>

Done.

---

*[explain.md, line 172 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1073#-KcUY5MdxbsxUm-gsKXK:-KcYrs1E9OJCsAqNxuds:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cf30b6dea9e262624e9e3027c20886daffcf16b3/explain.md#L172)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

nit: `In some cases, the <strong>Ordering</strong>` details...
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1073)*

**Comment by knz** [comment]
> Can you help me in coming up with a seemingly reasonable query that can be optimized by understanding that the query planner splitting things into levels, and then what that optimization would be?

I'm blanking on this one. I do know that explain is used to do that but I think I've never really done it myself because I never found myself in the role of a db user. perhaps we can forward the question to the team.

**Comment by knz** [comment]
Suggestion from @bdarnell: use EXPLAIN on a simple query on a table that does not have an index on the queried columns. Add an index. Then show that EXPLAIN on the same query does not use a full table scan any more.

**Comment by sploiselle** [comment]
@knz Great. Adding these as follow-on items in #1086

**Pull request merged**